### ISSUE TITLE
Automatic list update not ok from the home screen (ok from the drawer)

### ISSUE THREAD

**Issue by falzonv** [falzonv/discreet-launcher#13]
This ticket is created following ticket #10 opened by @xi7

In the second demo video, an application is installed from the home screen (opening the file manager in the favorites) but the applications list is not updated.

Quote from myself:

> I think this is because the events receiver is located with the complete applications list (because most of the time users will open their store from this list or uninstall an application which is not in their favorites).
This afternoon I made a dozen of install/remove to test and they were all working fine but none was launch from the home screen (each time I first opened F-Droid from the complete list), interesting test and indeed it doesn't work in this case.
>
> I will correct this (created ticket #13 for follow-up) but it will probably not be before a few days because I do this on my spare time and next week will be quite busy.
Hopefully this is not a major bug as it should work normally from the drawer and there is always the refresh button (see below).

**Comment by falzonv** [comment]
This bug has been fixed in release v2.1.0 which is available on F-Droid since today.

**Comment by xi7** [comment]
https://user-images.githubusercontent.com/67331785/120175486-a7603c80-c1dc-11eb-927d-6b3eb31664bd.mp4

auto refresh when enable / disable apps.

and maybe u can delete this function after solve it (refresh the apps list) obsolete function.

**Comment by falzonv** [comment]
Indeed, there is no auto update in this case because I cannot detect when the user disable/enable apps.

When applications are installed/removed, internal messages are sent by the Android system (named PACKAGE_ADDED and PACKAGE_REMOVED) and I can receive them and trigger an update of the applications list.

But there is no message when an application is disabled (or enabled) so I cannot detect that an update would be needed.
It is the main reason why I kept this manual update feature (but as you may have noticed, it is now much lower in the menu).

Unfortunately I will not be able to solve this due to the way it is managed in Android.
The only solution would be to update the applications list each time the user interacts with the launcher (open favorites, open drawer, open notification, etc.) but this would waste a lot of resources because the list would be constantly reloaded.

**Comment by xi7** [comment]
https://apkpure.com/br/lite-launcher/com.ck.dev.litelauncher

Keikai Launcher

work very well when enable/disable systems app and auto refresh list, so u can delete menu refresh.

**Comment by falzonv** [comment]
Hello,

Thank you for the suggestion!

I tried the Light Android Launcher and indeed the enable/disable works fine (by the way this minimalist app is quite beautiful).

Looking at the source code, in fact it does what I described above: there is no auto-refresh, the complete list of applications is retrieved from the system each time the user opens the applications drawer.

> The only solution would be to update the applications list each time the user interacts with the launcher (open favorites, open drawer, open notification, etc.) but this would waste a lot of resources because the list would be constantly reloaded.

In the case of Light Android Launcher, there is not much workload because it displays only the applications names (doesn't need to load and prepare all the icons which is the most intensive part of the process).

For Discreet Launcher, my goal is that when the user swipes to open the applications drawer, all the applications *with their icons* are *immediately* displayed (at least, as soon as technically possible), so rebuild the list each time is not an option (too slow).

Also, only pre-installed applications can be disabled and most users will not do that often. For example they will disable most bloatware apps after purchasing the phone, but may not need to disable anything else before months or years if their "disabling session" was complete enough.
If the list was updated each time the drawer is opened, resources would always be spent (and delay introduced) even if the user never disable any app in the future.

Best regards.
